### DISCUSSION TITLE
How to write facts based on other facts

### DISCUSSION THREAD

**Discussion by vitrun** [pythological/kanren#67] [Q&A] [answered]
The following code doesn't work as expected because two invocation of `Line('B', 'C')` result in different goal functions.

```python
Line = Relation()
LengthOf = Relation()

facts(LengthOf, (Line('B', 'C'), 1))
l = var()
run(1, l, LengthOf(Line('B', 'C'), l))
```

Though possible, the workaround is not elegant:
```python
g = Line('B', 'C')
facts(LengthOf, (g, 1))
l = var()
run(1, l, LengthOf(g, l))
```

What's the best practice to state facts based on other facts?

**Comment by brandonwillard** [comment, marked as answer]
`Line(*args)` should return a _goal_ object, which is generally a callable that takes miniKanren state objects (i.e. "substitution"/unification mappings) and returns a stream of new states.  Those goal objects are chained together using the `ldisj` and `lconj` operators (e.g. via the helper functions/aliases `conde`, `lall`, `land`, etc.) and evaluated by `run`.  

Altogether, this means that the expression `LengthOf(Line(*args), ...)` will create a goal that relates a goal to something else; in other words, the result is some sort of "higher-order" relation on the machinery of miniKanren itself (i.e. goals).  That's an interesting path to go down, but it might not be what you want.

More specifically, the goal generated by `LengthOf(Line(*args), ...)` will attempt to unify the two distinct goal instances created by the expressions `Line("B", "C")`, and that's where it probably fails.  If you really do want this to work, you'll need to subclass `Relation` and define equality between your subclasses and/or specify how instances of those subclasses unify (e.g. create `_unify` and `_reify` implementations via [`logical-unification`](https://github.com/pythological/unification/)).
